### ISSUE TITLE
Rendering KMW tiles should depend on associated modules

### ISSUE THREAD

**Issue by jimmymadon** [google/site-kit-wp#7337]
## Feature Description

Every individual key metric widget tile depends on one or more modules as determined by the `modules` option when registering the widget tile. We need to guard against rendering tiles when their associated modules are not active. All other widgets on our dashboard currently use the `whenActive` HOC to achieve this.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- Key Metric widget tiles which rely on their corresponding module should not be rendered when their module is disconnected. 
    - As of this issue, we have 1 Search Console widget tile and the remaining all rely on GA4. 
    - Note: In the future, there will be tiles which rely on the GA4 - AdSense integration and while this tile is not to be tested now, solutions for implementing this issue should bear this in mind. 
- As part of #6263 and #6265, we implemented both, an individual "Connect GA4 CTA" tile and a larger full-width "CTA banner" for when GA4 is disconnected after KMW setup. To simplify the engineering/implementation and using the existing Widgets API rendering a "fallback" component, the behaviour of those CTAs when GA4 is disconnected after KMW setup should be updated as follows:
    - The “Connect GA” individual metric tile from #6263 should "stretch" taking the same amount of space that would have been otherwise taken by tiles reliant on GA4. This could be either 1 (quarter), 2 (half), 3 (three-fourth) or 4 (full-width) space.
    - If the "Top Performing Keywords" tile which relies on Search Console is rendered, it should be rendered first in the row of metrics. This will ensure that only a single GA4 CTA tile will be rendered "continuously" for the remainder spaces.
    - The full width "Connect GA CTA Banner" should not be rendered anymore. If all selected metrics rely on GA4, we should render this component design as part of a future issue (#7278 will be updated to handle this case).
    - The AC of #7336 should still be respected, i.e. the "+ Add a metric" widget tiles should be added appropriately at the end of the row if the number of metrics selected is 2 or 3. (This point is here for completeness / regression testing and nothing new needs to be added as part of this issue.)
- There should be no network requests from widgets that are not rendered and any console errors originating from such requests.

## Implementation Brief

NOTE TO IB Reviewer: Please review ACs above before checking the IB below.

* A [POC branch](https://github.com/google/site-kit-wp/compare/poc/kmw-when-active) has been created to guide the approach that is highlighted below.
* The general idea here is to reuse the Widgets API functionality as much as possible where each widget is responsible for its own rendering and knows little about other widgets. 
* We should use the `whenActive` HOC to render all the new KM tile widgets passing in the module that they depend on. This will ensure that `WidgetNull` is rendered instead of the widget components when the associated module is disconnected.
* The `whenActive` HOC allows us to pass the `FallbackComponent` in the event that the above mentioned module is not active. For all GA4 reliant modules, we should pass the `ConnectGA4CTATileWidget` created in #6263. This component will be rendered as a fall back automatically. This fallback component will be rendered with the same props that the widget would normally receive. 
* The more imperative logic which checks for the number of GA4 metric tiles selected within `ConnectGA4CTATileWidget` should be removed. 
* The `useWidgetStateEffect` should be used within this fallback component to render an `OverrideComponent`. The `OverrideComponent` here would be the generic `ConnectModuleCTATile`. This will be used automatically to combine all GA4 widgets which have the `WidgetNull` state so as to stretch the fall back component appropriately.  However, the component type provided to `useWidgetStateEffect` is not rendered with any props and is wrapped with a generic `Widget` automatically. So `ConnectModuleCTATile` needs to be updated to not depend on any props (which is doable as the POC shows).

### Test Coverage

* No new tests required since we are using the existing tested Widgets API functions here.

### Connect GAA CTA

- Set up Site Kit with the `userInput` feature flag enabled and GA4 connected, and setup Key Metrics.
- Disconnect Analytics.
- Return to the Dashboard and verify that a single instance of the Connect GAA CTA widget is now rendered in place of the GA4 Key Metrics widgets (currently, these are all of them barring the "Top performing keywords" widget).
- Verify that the "Top performing keywords" widget appears first in the list of widgets. To show this widget ensure that the first Key Metrics question, "What is the main purpose of this site?", is answered with either "Sell products or services" or "Share a business card or portfolio".

### Connect AdSense CTA

- Set up Site Kit with the `userInput` feature flag enabled, connect GA4 and AdSense, and setup Key Metrics.
- Use the snippets below to show the skeleton version of the Top Earning Content widget.
- Disconnect AdSense.
- Return to the Dashboard and verify that the Connect AdSense CTA widget is now rendered in place of the Top Earning Content widget.
- Now disconnect Analytics.
- Return to the Dashboard and verify that the Connect GA4 CTA widget is now rendered in place of the Top Earning Content widget.

```js
// Run this snippet to show the Top Earning Content widget.
googlesitekit.data.dispatch( 'core/user' ).receiveGetKeyMetricsSettings( {
    widgetSlugs: [ 'kmAnalyticsAdSenseTopEarningContent' ]
} );

// Combine it with other widgets to further test the layouts. Bear in mind that the order of the widgets is not dependent on the order of the `widgetSlugs` array, but rather is down to the order and configuration of the widget registration.
googlesitekit.data.dispatch( 'core/user' ).receiveGetKeyMetricsSettings( {
    widgetSlugs: [ 'kmAnalyticsAdSenseTopEarningContent', 'kmAnalyticsEngagedTrafficSource', 'kmSearchConsolePopularKeywords', 'kmAnalyticsTopTrafficSource' ]
} );
```

## Changelog entry

* Update key metric widget rendering for consistency with other widgets regarding module dependencies.

**Comment by aaemnnosttv** [comment]
Thanks @jimmymadon – the IB looks good for the most part. A few points that are done in the POC that are worth highlighting:

- The fallback component used by `whenActive` is rendered with all of the same props that the widget would normally receive
- The "override component" as it's called internally, which is what is rendered in the case of combined states, is not rendered with any props and is wrapped with a generic `Widget` automatically
    -  This is important because the component type provided to `useWidgetStateEffect` is what will be used as the override component, so this needs to be updated to not depend on any props (which is doable as the POC shows)

Aside from that, the test coverage section is empty but you added a related bullet above. Let's be explicit if we expect to need to add/update any tests. It can be assumed that there should be no failing tests or broken VRT :)

**Comment by aaemnnosttv** [comment]
Great, thanks @jimmymadon 👍 

IB ✅

**Comment by techanvil** [comment]
I've moved this back to Execution pending the outcome of https://github.com/google/site-kit-wp/issues/6258#issuecomment-1669489810.

**Comment by mohitwp** [comment]
@techanvil FYI. Seems like AC is implemented only for widget displaying through tailor widgets. If user pick widgets then design is not as per AC. In both cases, widgets should display as per AC.

cc @jimmymadon @wpdarren 

**When user pick metrics**

![image](https://github.com/google/site-kit-wp/assets/94359491/2504358a-3b64-4ff8-82c1-8d98533e483f)


**In case of tailored metrics**

![image](https://github.com/google/site-kit-wp/assets/94359491/51676c61-bb7a-47a3-b584-509dcf7dfaa2)

**Comment by mohitwp** [comment]
Thanks for the update @techanvil !

**Comment by techanvil** [comment]
Hi @mohitwp, thanks for raising this. It was a mistake in the QAB - in fact, the Analytics CTA is what takes precedence in the case where both AdSense and GA4 are disconnected.

I've updated the QAB to reverse the order of module disconnection. Please take a look and give this another run.

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Verified If the "Top Performing Keywords" tile which relies on Search Console is rendered, it renders first in the row of metrics.
- Verified KMW tiles appears as per status of associated modules as per AC. 

**Note** 
**1)** Add a metric widget will be implemented by #7336 .
**2)** Full width Connect GA4 banner will be updated by #7278.

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/f993eb06-4f26-4d92-8339-c59f2d34c387)

![image](https://github.com/google/site-kit-wp/assets/94359491/b4bba477-8a42-4f2d-b210-71741093165f)

![image](https://github.com/google/site-kit-wp/assets/94359491/d02083bd-a729-438c-9c8a-4c4b7a907f8f)

![image](https://github.com/google/site-kit-wp/assets/94359491/e118b20e-e9d5-47c8-90a3-0b255d126a65)

![image](https://github.com/google/site-kit-wp/assets/94359491/2ee09e4e-9095-41dd-9c96-74bf584818a9)

![image](https://github.com/google/site-kit-wp/assets/94359491/0369636b-6dad-44a3-be06-7ccce56a94a9)

</details>
